rfd/0207-vnet-ssh.md

```diff
@@ -0,0 +1,392 @@
+```
+
+Because the user may log in or out of clusters and new DNS zones may be added,
+this line may need to be updated during the lifetime of the VNet process.
+To handle this, VNet will check for this entry each time it successfully
+resolves a DNS query for an SSH server.
+If it is missing or does not contain an entry that matches the current query, it
+will update the entry.
+
+The trailing comment `# Teleport VNet` is used so that VNet can reference and
+update a single line in the `known_hosts` file.
+Before shutdown, VNet will delete the entry.
+If the process is killed without being able to delete the entry, it will be
+overwritten the next time VNet starts, this should not cause any problems.
+
+#### Search domains
+
+One possibility to shorten the DNS names for SSH hosts, e.g. so the user
+could type `ssh <user>@<hostname>` instead of `<user>@<hostname>.ssh.<proxy address>`,
+would be to configure a DNS [search domain](https://en.wikipedia.org/wiki/Search_domain)
+in the host OS.
+For example, if your proxy address is `teleport.example.com` you could configure
+a search domain for `ssh.teleport.example.com.`.
+The OS DNS resolver will then reference that search domain whenever it receives
+a query for a DNS name that is not fully-qualified (does not end with a `.`).
+So if the user types `ssh user@hostname`, the OS DNS resolver would
+fully-qualify that to `hostname.ssh.teleport.example.com`, which is the query
+that would be sent to the VNet DNS nameserver, and everything should work.
+
+This is something we could potentially configure automatically for the user, but
+I think it might be likely to conflict with many corporate internal network
+search domains, so I don't plan to implement this in VNet.
+Another downside to this approach is that VNet would probably receive queries
+for almost all DNS queries on the host, because users usually don't fully-qualify their
+domain names.
+E.g. if the user tries to look up `google.com` VNet would get a DNS query for
+`google.com.ssh.teleport.example.com.`.
+When the VNet name lookup fails the OS resolver should properly handle the
+lookup for `google.com.`, but it may slow things down and add unnecessary load to
+the VNet process which may also generate queries to the cluster.
+
+#### DNS resolution
+
+When the VNet process receives a DNS query for `<hostname>.ssh.<namespace>`,
+this is how it will be resolved:
+
+1. If it has already resolved this name and assigned a VNet IP address for the
+   host, it will return the assigned IP address.
+1. If the name matches `*.ssh.<proxy-address>` or `*.ssh.<custom-dns-zone>`:
+   1. VNet will determine which clusters the proxy address or custom DNS zone
+      matches. If none, the DNS request is forwarded upstream.
+   1. VNet will apply any matching proxy templates found in the user's
+      `TELEPORT_HOME`.
```

Comment:
This might be a bit of a pain for some users, as I'm sure that people usually configure tsh through `~/.tsh/config/config.yaml` which Teleport Connect does not use. But it is something we might tackle in Q2, so it's not _that_ big of a problem.

Reply:
Yeah I don't think proxy templates are going to be that common to use for this, but I know there are some specific customers that use them heavily, and I wanted to make it at least possible to use them from connect, but idk using TELEPORT_HOME seemed better than the default tsh home